File: src/Ivy.Docs.Tools/rust_cli/src/link_converter.rs
use regex::Regex;
use std::collections::HashSet;
use crate::utils;

pub struct LinkConverter {
    current_file_path: String,
    link_regex: Regex,
}

impl LinkConverter {
    pub fn new(current_file_path: &str) -> Self {
        Self {
            current_file_path: current_file_path.to_string(),
            link_regex: Regex::new(r"!?\[(.*?)\]\((.*?)\)").unwrap(),
        }
    }

    pub fn convert(&self, markdown: &str) -> (HashSet<String>, String) {
        let mut types = HashSet::new();
        
        let mut result = markdown.to_string();
        
        // Find all matches first to avoid recursive replacement issues
        let matches: Vec<(String, String, String)> = self.link_regex.captures_iter(markdown)
            .filter(|cap| !cap[0].starts_with("!"))
            .map(|cap| (cap[0].to_string(), cap[1].to_string(), cap[2].to_string()))
            .collect();
            
        for (full_match, text, link) in matches {
            if link.starts_with("app://") || link.starts_with("http://") || link.starts_with("https://") ||
               link.starts_with("mailto:") || link.starts_with("tel:") || link.starts_with("#") {
                continue;
            }

            let fragment_index = link.find('#');
            let (link_without_fragment, fragment) = match fragment_index {
                Some(i) => (&link[..i], &link[i..]),
                None => (link.as_str(), ""),
            };

            let path = utils::get_path_for_link(&self.current_file_path, link_without_fragment);
            let type_name = utils::get_type_name_from_path(&path);
            let app_id = utils::get_app_id_from_type_name(&type_name);

            types.insert(type_name.clone());
            let replacement = format!("[{}](app://{}{})", text, app_id, fragment);
            result = result.replace(&full_match, &replacement);
        }

        (types, result)
    }
}

File: src/Ivy.Docs.Tools/LinkConverter.cs
using System.Text.RegularExpressions;

namespace Ivy.Docs.Tools;

public partial class LinkConverter(string currentFilePath)
{
    public (HashSet<string> types, string markdown) Convert(string markdown)
    {
        var types = new HashSet<string>();
        var regex = MyRegex();
        var matches = regex.Matches(markdown);
        foreach (Match match in matches)
        {
            if (match.Value.StartsWith("!")) continue; //this is an image
            var (type, linkReplacement) = ConvertLink(match);
            if (type == null) continue;
            types.Add(type);
            markdown = markdown.Replace(match.Value, linkReplacement);
        }

        return (types, markdown);
    }

    private (string? type, string? linkReplacement) ConvertLink(Match match)
    {
        var text = match.Groups[1].Value;
        var link = match.Groups[2].Value;

        if (link.StartsWith("app://") || link.StartsWith("http://") || link.StartsWith("https://") ||
            link.StartsWith("mailto:") || link.StartsWith("tel:") || link.StartsWith("#"))
        {
            return (null, null); //do nothing
        }

        // Strip fragment identifier (#...) from link before processing
        var fragmentIndex = link.IndexOf('#');
        var linkWithoutFragment = fragmentIndex >= 0 ? link[..fragmentIndex] : link;
        var fragment = fragmentIndex >= 0 ? link[fragmentIndex..] : "";

        var path = Utils.GetPathForLink(currentFilePath, linkWithoutFragment);
        var type = Utils.GetTypeNameFromPath(path);
        var appId = Utils.GetAppIdFromTypeName(type);

        return (type, $"[{text}](app://{appId}{fragment})");
    }

    [GeneratedRegex(@"!?\[(.*?)\]\((.*?)\)")]
    private static partial Regex MyRegex();
}
